qlib/VscDebugAdapter.qm

```diff
@@ -0,0 +1,1248 @@
+                                #if (len < 0 ) {
+                                    #throw "VDA-ERROR", "Missing content length";
+                                #}
+                                payload = len > 0;
+                            } else {
+                                string s = buff.toString("UTF-8");
+                                # read headers
+                                *list l = s =~ x/^([A-Za-z\-]+):(.+)$/;
+                                if (l[0] == "Content-Length") {
+                                    trim l[1];
+                                    len = int(l[1]);
+                                    log(DUV_DEBUG_1, "Content-Length: %d", len);
+                                }
+                            }
+                            remove buff;
+                        } else if (c != <0D>) {
+                            buff += c;
+                        }
+                    }
+                    while (m_queue.size() > 0) {
+                        any msg = m_queue.get();
+                        try {
+                            # event from debug server
+                            if (msg.type == "event" ) {
+                                processDebugEvent(msg);
+                            }
+                            if (msg.type == "vsc_output") {
+                                hash h.category = msg.category;
+                                h.output = msg.output;
+                                stderr.printf("vscOutput: %y\n", h);
+                                sendVscEvent("output", h);
+                            }
+                        } catch (hash<ExceptionInfo> ex) {
+                            log(DUV_INFO, "ex: %y", ex);
+                        }
+                    }
+                }
+                sendVscEvent("terminated");
+            } catch (hash<ExceptionInfo> ex) {
+                log(DUV_INFO, "ex: %y", ex);
+                rethrow;
+            }
+        }
+
+        #! execute command (and exit)
+        public execute(string cmd) {
+            auto res;
+            switch (cmd) {
+                case 'pgmlist':
+                    res = doCommand(('cmd': 'program//list', 'listFiles': True));
+                    break;
+                default:
+                    throw "VDA-ERROR", sprintf("Unknown command: %y", cmd);
+            }
+            stdout.write(make_json(res)+"\r\n");
+        }
+
+        private:internal auto resolveValue(string name, *string val, hash strings, auto defVal) {
+            if (exists val) {
+                if (!exists strings{val}) {
+                    throw "VDA-ERROR", sprintf("%y value %y is not found in %y", name, val, keys strings);
+                }
+                return strings{val};
+            } else {
+                return defVal;
+            }
+        }
+
+        #! process event from debugger
+        private:internal processDebugEvent(hash event) {
```

Comment:
this method is also very long and complex, also using flags as ints not named constants, it makes it really hard to understand, can u pls update it?

Reply:
It requires another issue to introduce constants in DebugProgramControl.qm and update all stuff dependent on library.